ompi/mca/coll/solo/coll_solo_barrier.c

```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright (c) 2019      The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "coll_solo.h"
+int mac_coll_solo_barrier_intra(struct ompi_communicator_t *comm, mca_coll_base_module_t * module)
+{
+    mca_coll_solo_module_t *solo_module = (mca_coll_solo_module_t *) module;
+
+    /* Enable solo module if necessary */
+    if (!solo_module->enabled) {
+        mca_coll_solo_lazy_enable(module, comm);
+    }
+
+    int rank = ompi_comm_rank(comm);
+    /* Atomic add to current ctrl_buf */
+    char *barrier_ctrl_bufs = solo_module->ctrl_bufs[0] + opal_cache_line_size;
+    opal_atomic_add_fetch_32((opal_atomic_int32_t *) (barrier_ctrl_bufs + solo_module->barrier_tag * opal_cache_line_size), 1);
+    while (*((int32_t *) (barrier_ctrl_bufs + (solo_module->barrier_tag) * opal_cache_line_size)) != ompi_comm_size(comm)) {
+        opal_progress();
+    }
+
+    /* Set previous used ctrl_buf to 0 */
+    if (rank == 0) {
```

Comment:
This is not safe as process rank 0 can be delayed for external reasons and the rest of the local processes might start a new barrier, colliding with the previous one.

Reply:
I address this issue by having three buffers instead of two in the ctrl_buf for MPI_Barrier. 
For each barrier, the next, the current and the previous will always use different buffers.